http4s/src/main/scala/Http4sConsul.scala

```diff
@@ -62,20 +66,118 @@ final class Http4sConsulClient[F[_]](
   def addCreds(req: Request[F]): Request[F] =
     credentials.fold(req){case (un,pw) => req.putHeaders(Authorization(BasicCredentials(un,pw)))}
 
-  def kvGet(key: Key): F[Option[String]] = {
+  /** A nice place to store the Consul response headers so we can pass them around */
+  case class ConsulHeaders(
+    index:       Long,
+    lastContact: Long,
+    knownLeader: Boolean
+  )
+
+  /** Helper function to get the value of a header out of a Response */
+  def extractHeaderValue(header: String, response: Response[F]): F[String] = {
+    response.headers.get(header.ci) match {
+      case Some(header) => F.pure(header.value)
+      case None         => F.pure(new RuntimeException(s"Header not present in response: $header")).flatMap(F.raiseError)
+    }
+  }
+
+  /** Helper function to get Consul GET request metadata from response headers */
+  def extractConsulHeaders(response: Response[F]): F[ConsulHeaders] = {
+    for {
+      index       <- extractHeaderValue("X-Consul-Index", response).map(_.toLong)
+      lastContact <- extractHeaderValue("X-Consul-LastContact", response).map(_.toLong)
+      knownLeader <- extractHeaderValue("X-Consul-KnownLeader", response).map(_.toBoolean)
+    } yield ConsulHeaders(index, lastContact, knownLeader)
+  }
+
+  /**
+    * Encapsulates the functionality for parsing out the Consul headers from the HTTP response and decoding the JSON body.
+    * Note: these headers are only present for a portion of the API.
+    */
+  def extractQueryResponse[A](response: Response[F])(implicit d: EntityDecoder[F, A]): F[QueryResponse[A]] = response match {
+    case Successful(r) =>
+      for {
+        headers     <- extractConsulHeaders(response)
+        decodedBody <- d.decode(r, strict = false).fold(throw _, identity)
+      } yield {
+        QueryResponse(decodedBody, headers.index, headers.knownLeader, headers.lastContact)
+      }
+    case failedResponse =>
+      F.pure(UnexpectedStatus(failedResponse.status)).flatMap(F.raiseError)
```

Comment:
Can be another direct `F.raiseError`.

Since we're speaking to a specific service, rather than as a general client, we can probably safely grab the error message as a String or JSON for debugging purposes.  http4s' use of `UnexpectedStatus` is that it's hard to make assumptions in the general case.